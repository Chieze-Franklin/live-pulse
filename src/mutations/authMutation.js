import { gql } from 'apollo-boost';

export const SIGNUP_MUTATION = gql `
  mutation signup ($email: String!, $password: String!, $data: UserUpdateInput!) {
    signUp (email:$email, password:$password, data: $data) {
      token,
      user {
        email,
        displayName,
        updatedAt,
        dashboards {
          icon,
          publicUrl,
          title
        }
      }
    }
  }
`

export const SIGNIN_MUTATION = gql`
  mutation signin ($email: String!, $password: String!) {
    signIn (email:$email, password:$password) {
      token,
      user {
        email,
        displayName,
        dashboards {
          icon,
          publicUrl,
          title
        }
        updatedAt
      }
    }
  }
`
