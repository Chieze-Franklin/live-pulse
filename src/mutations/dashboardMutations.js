import gql from "graphql-tag";

/**
 * hanles dashboard creating mutation to the server
 * @param title [string] - required, to be suplied to the mutation handler
 * @param updateInterval [integer]
 * @param icon [string]
 * 
 * @returns created dashboard
 */
export const DASHBOARD_MUTATION = gql`
  mutation createDashboard ($title: String!, $upInt: Int, $icon: String) {
    createDashboard (data: {
      title:$title
      updateInterval:$upInt
      icon:$icon
    })
    {
      id,
      title,
      updateInterval,
      updatedAt
    }
  }
`
