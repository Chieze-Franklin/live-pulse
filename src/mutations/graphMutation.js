import gql from "graphql-tag";

/**
 * Update graph's general properties
 */
export const UPDATE_GRAPH_GENERAL_MUTATION = gql`
  mutation updateGeneral($id:ID!, $data:GraphUpdateInput!) {
    updateGraph(id:$id, data:$data) {
      id,
      title,
      updateInterval,
      xAxisLabel,
      yAxisLabel
    }
  }
`;

/**
 * Updates graph's Viaribles properties
 */
export const UPDATE_GRAPH_VARIABLES_MUTATION = gql`
mutation updateVariables($id:ID!, $data:GraphUpdateInput!) {
  updateGraph(id:$id, data:$data) {
    variables
  }
}
`;

/**
 * Update graph's LineGenerator properties
 */
// export const UPDATE_GRAPH_LINE_G_MUTATION = gql`
//   mutation 
// `;
