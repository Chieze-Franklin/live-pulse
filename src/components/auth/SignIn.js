import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import Checkbox from '@material-ui/core/Checkbox';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import withStyles from '@material-ui/core/styles/withStyles';
import { styles } from './auth.styles';
import { Mutation } from 'react-apollo';
import { SIGNIN_MUTATION, SIGNUP_MUTATION } from '../../mutations/authMutation';
import { AUTH_TOKEN, UID, U_EMAIL } from '../../constants';

const authToken = localStorage.getItem(AUTH_TOKEN);
const uid = localStorage.getItem(UID);
const uemail = localStorage.getItem(U_EMAIL);

const SignIn = (props) => {
  const { classes } =  props;
  const [login, setLogin] = useState(true);
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmEmail, setConfirmEmail] = useState('');

  //bruteforce fix for authentication redirecting
  if (authToken && uid && uemail) {
    window.location.href = '/dashboard'
    return (<div>redirecting</div>);
  }

  return (
    <main className={classes.main}>
      <CssBaseline />
      <Paper className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h5">
          {login ? 'Log In' : 'Create a live pulse account'}
        </Typography>
        <div className={classes.form}>
          <FormControl margin="normal" required fullWidth>
            <InputLabel htmlFor="email">Email Address</InputLabel>
            <Input type="email" id="email" name="email" autoComplete="email" autoFocus 
              value={email}
              onChange={e => setEmail(e.target.value)}
            />
          </FormControl>
          <FormControl margin="normal" required fullWidth>
            <InputLabel htmlFor="password">Password</InputLabel>
            <Input name="password" type="password" id="password" autoComplete="current-password" 
              value={password}
              onChange={e => setPassword(e.target.value)}
            />
          </FormControl>
          {!login && (
            <FormControl margin="normal" required fullWidth>
              <InputLabel htmlFor="email">Confirm Email Address</InputLabel>
              <Input type="email" id="email2" name="email2" autoFocus 
              value={confirmEmail}
              onChange={e => setConfirmEmail(e.target.value)}
              />
            </FormControl>
          )}

          <FormControlLabel
            control={<Checkbox value="remember" color="primary" />}
            label="Remember me"
          />
          <p onClick={() => setLogin(!login)} >
            {login
              ? `Don't have an account? create one`
              : `Already have an account? log in`
            }

          </p>
          <div className="mutation-button">
            <Mutation
              mutation={login ? SIGNIN_MUTATION : SIGNUP_MUTATION}
              variables={{ email, password, data:{ email: confirmEmail}}}
              onCompleted={data => authenticate(data, login)}
              onError={error => handleError(error)}
            >
              {authMutation => (
                <Button
                  type="submit"
                  fullWidth
                  variant="contained"
                  color="primary"
                  className={classes.submit}
                  onClick={authMutation}
                >
                  {login ? 'Log in' : 'Create account'}
                </Button>
              )}
            </Mutation>
          </div>
        </div>
      </Paper>
    </main>
  );
}

/**
 * authenticate function to handle signup & signin mutation
 */
const authenticate = async (data, login) => {
  console.log(data);
  const { token, user:{ id, email} } = (login ? data.signIn : data.signUp);
  localStorage.setItem(AUTH_TOKEN, token);
  localStorage.setItem(UID, id);
  localStorage.setItem(U_EMAIL, email);
  window.location.href = '/dashboard'; // react-router-redirect failed to work so improvising for now
}

/**
 * handles error on failed login or signup attempts.
 * does nothing serious for nm=ow, just logs the error to console 
 * for debugging purposes.
 */
const handleError = async error => {
  console.log(error);
}

SignIn.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(SignIn);
