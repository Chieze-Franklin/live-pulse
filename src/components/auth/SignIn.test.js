import React from 'react';
import  renderer from  'react-test-renderer';
import  { act } from  'react-test-renderer';
import { MockedProvider } from 'react-apollo/test-utils';
import SignIn from './SignIn';
import { SIGNIN_MUTATION } from '../../mutations/authMutation';
import wait from 'waait';

describe('Signin/SignUp Component ', () => {
  const signIn = { token: 'fsdgfgfhghl', user: {id:'hnhgdc', emai:'wale@mail.com'}}
  const mocks = [
    {
      request: {
        query: SIGNIN_MUTATION,
        variables: {
          email: 'wale@email.com',
          password: '1234',
        },
      },
      result: {
        data: { signIn }
      }
    },
  ];
  const component = renderer.create(
    <MockedProvider mocks={mocks} addTypename={false}>
      <SignIn />
    </MockedProvider>
   );
   const compInstance = component.root;
   const button = compInstance.findByProps({type: 'submit'});

  it('Regression test => Should match snapshot', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Should render email field correctly', () => {
    expect(compInstance.findByProps({name: 'email'}));
  });
  it('Should render password field correctly', () => {
    expect(compInstance.findByProps({name: 'password'}));
  });
  it('Should render SignIN button correctly', () => {
    expect(button.props.children).toEqual('Log in');
  });
  it('Should signIN a user successfully', async () => {  
    act(() => {
      button.props.onClick();
    });
    await wait(0);
    expect(compInstance.findByProps({id: 'errorMsg'}).props.children).toEqual('.')

  });
  it('Shows Signup components when create account is clicked', () => {
    const p = compInstance.findByProps({id: "acct-auth"});
    act(() => {
      p.props.onClick();
    });
    expect(p.props.children).toBe('Already have an account? log in');
    expect(compInstance.findByProps({name: 'email2'}));
    expect(compInstance.findByProps({type: 'submit'})
      .props.children).toEqual('Create account');
  });
  it('Authenticate function should setItems in localStorage', () => {
    
  });
});
