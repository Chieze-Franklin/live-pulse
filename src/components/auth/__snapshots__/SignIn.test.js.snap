// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signin/SignUp Component  Regression test => Should match snapshot 1`] = `
<main
  className="SignIn-main-1"
>
  <div
    className="MuiPaper-root-6 MuiPaper-elevation2-10 MuiPaper-rounded-7 SignIn-paper-2"
  >
    <div
      className="MuiAvatar-root-33 MuiAvatar-colorDefault-34 SignIn-avatar-3"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-36"
        color={undefined}
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <g
          fill="none"
        >
          <path
            d="M0 0h24v24H0V0z"
          />
          <path
            d="M0 0h24v24H0V0z"
            opacity=".87"
          />
        </g>
        <path
          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
        />
      </svg>
    </div>
    <p
      id="errorMsg"
    >
      
    </p>
    <h1
      className="MuiTypography-root-45 MuiTypography-h5-61"
    >
      Log In
    </h1>
    <div
      className="SignIn-form-4"
    >
      <div
        className="MuiFormControl-root-81 MuiFormControl-marginNormal-82 MuiFormControl-fullWidth-84"
      >
        <label
          className="MuiFormLabel-root-96 MuiFormLabel-required-101 MuiInputLabel-required-89 MuiInputLabel-root-85 MuiInputLabel-formControl-90 MuiInputLabel-animated-93"
          data-shrink={false}
          htmlFor="email"
        >
          Email Address
          <span
            className="MuiFormLabel-asterisk-102"
          >
             *
          </span>
        </label>
        <div
          className="MuiInputBase-root-116 MuiInput-root-103 MuiInput-underline-107 MuiInputBase-formControl-117 MuiInput-formControl-104"
          onClick={[Function]}
        >
          <input
            aria-describedby={undefined}
            aria-invalid={false}
            autoComplete="email"
            autoFocus={true}
            className="MuiInputBase-input-126 MuiInput-input-111 MuiInputBase-inputType-129 MuiInput-inputType-114"
            defaultValue={undefined}
            disabled={false}
            id="email"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={true}
            rows={undefined}
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-81 MuiFormControl-marginNormal-82 MuiFormControl-fullWidth-84"
      >
        <label
          className="MuiFormLabel-root-96 MuiFormLabel-required-101 MuiInputLabel-required-89 MuiInputLabel-root-85 MuiInputLabel-formControl-90 MuiInputLabel-animated-93"
          data-shrink={false}
          htmlFor="password"
        >
          Password
          <span
            className="MuiFormLabel-asterisk-102"
          >
             *
          </span>
        </label>
        <div
          className="MuiInputBase-root-116 MuiInput-root-103 MuiInput-underline-107 MuiInputBase-formControl-117 MuiInput-formControl-104"
          onClick={[Function]}
        >
          <input
            aria-describedby={undefined}
            aria-invalid={false}
            autoComplete="current-password"
            autoFocus={undefined}
            className="MuiInputBase-input-126 MuiInput-input-111 MuiInputBase-inputType-129 MuiInput-inputType-114"
            defaultValue={undefined}
            disabled={false}
            id="password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={true}
            rows={undefined}
            type="password"
            value=""
          />
        </div>
      </div>
      <label
        className="MuiFormControlLabel-root-133"
      >
        <span
          className="MuiButtonBase-root-155 MuiIconButton-root-149 MuiPrivateSwitchBase-root-145 MuiCheckbox-root-139 MuiCheckbox-colorPrimary-143"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role={undefined}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label-154"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-36"
              color={undefined}
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <input
              autoFocus={undefined}
              checked={undefined}
              className="MuiPrivateSwitchBase-input-148"
              data-indeterminate={false}
              defaultChecked={undefined}
              disabled={undefined}
              id={undefined}
              name={undefined}
              onChange={[Function]}
              readOnly={undefined}
              required={undefined}
              tabIndex={undefined}
              type="checkbox"
              value="remember"
            />
          </span>
          <span
            className="MuiTouchRipple-root-184"
          />
        </span>
        <span
          className="MuiTypography-root-45 MuiTypography-body1-54 MuiFormControlLabel-label-138"
        >
          Remember me
        </span>
      </label>
      <p
        id="acct-auth"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        Don't have an account? create one
      </p>
      <div
        className="mutation-button"
      >
        <button
          className="MuiButtonBase-root-155 MuiButton-root-158 MuiButton-contained-169 MuiButton-containedPrimary-170 MuiButton-raised-172 MuiButton-raisedPrimary-173 MuiButton-fullWidth-183 SignIn-submit-5"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="submit"
        >
          <span
            className="MuiButton-label-159"
          >
            Log in
          </span>
          <span
            className="MuiTouchRipple-root-184"
          />
        </button>
      </div>
    </div>
  </div>
</main>
`;
